### PR TITLE
Harry/frontend refactor nochange

### PR DESCRIPTION
**Pull request by TuxedoFish**
## What does this PR do?
Refactored the marketplace, home page to use a consistent set of rows and columns and identical styling for easier re-use. Can now ensure regardless of how the aesthetic changes in uclapi.scss, the changes across pages will have a consistent feel.

Needs to be carried over into dashboard as in Nush's redesign.

## ✅ Pull Request checklist

- [ ] Is this code complete?
- [ ] Are tests done/modified?
- [ ] Are docs written for this PR? (if applicable)

## 🚨 Is this a breaking change for API clients?
Yes/No

## :squirrel: Deploy notes
For example: Need to run migrations as part of deployment

### CODE REVIEW

**Comment by TuxedoFish** [comment]
Just spotted issue with FAQ rendering will fix tomorrow

File: backend/uclapi/oauth/views.py [hunk 1]
```diff
@@ -1,761 +1,761 @@
...
+        return response
+
+    token.delete()
+
+    response = PrettyJsonResponse({
+        "ok": True,
+        "message": "App successfully deauthorised."
+    })
+    response.status_code = 200
+    return response
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/views.py]: Resolved in future commit

File: backend/uclapi/oauth/views.py [hunk 2]
```diff
@@ -1,761 +1,761 @@
...
-    return response
+import json
+import os
+import urllib.parse
+
+import redis
+from django.core import signing
+from django.core.serializers.json import DjangoJSONEncoder
+from django.core.signing import TimestampSigner
+from django.http import HttpResponse
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/views.py]: Resolved in future commit

File: backend/uclapi/oauth/tests.py [hunk 3]
```diff
@@ -1,817 +1,817 @@
...
+
+        content = json.loads(response.content.decode())
+        self.assertFalse(content["ok"])
+        self.assertEqual(
+            content["error"],
+            (
+                "The app with the Client ID provided does not have a "
+                "token for this user, so no action was taken."
+            )
+        )
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/tests.py]: Resolved in future commit

File: backend/uclapi/dashboard/views.py [hunk 4]
```diff
@@ -1,209 +1,214 @@
...
+        'initial_data': initial_data
+    })
+
+@ensure_csrf_cookie
+def about(request):
+    return render(request, 'AboutPage.html', {
+        'initial_data': {}
+    })
+
+@ensure_csrf_cookie
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/dashboard/views.py]: Resolved in future commit

File: backend/uclapi/dashboard/views.py [hunk 5]
```diff
@@ -1,209 +1,214 @@
...
+                "contact": app.webhook.contact
+            }
+        })
+
+    initial_data = json.dumps(user_meta, cls=DjangoJSONEncoder)
+    return render(request, 'Dashboard.html', {
+        'initial_data': initial_data
+    })
+
+@ensure_csrf_cookie
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/dashboard/views.py]: Resolved in future commit

File: backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py [hunk 6]
```diff
@@ -1,60 +1,60 @@
...
+            fields=[
+                ('id', models.AutoField(primary_key=True, serialize=False)),
+                ('setid', models.TextField(max_length=10)),
+                ('studentid', models.TextField(max_length=12)),
+                ('courseid', models.TextField(max_length=12)),
+                ('classgroupid', models.TextField(max_length=10)),
+                ('clsgrpcode', models.TextField(max_length=10)),
+                ('courseyear', models.BigIntegerField(blank=True, null=True)),
+                ('fixingrp', models.CharField(blank=True, max_length=1, null=True)),
+                ('inactive', models.CharField(blank=True, max_length=1, null=True)),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py]: Will not change migrations manually

File: backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py [hunk 7]
```diff
@@ -1,60 +1,60 @@
...
+            name='StuclassesB',
+            fields=[
+                ('id', models.AutoField(primary_key=True, serialize=False)),
+                ('setid', models.TextField(max_length=10)),
+                ('studentid', models.TextField(max_length=12)),
+                ('courseid', models.TextField(max_length=12)),
+                ('classgroupid', models.TextField(max_length=10)),
+                ('clsgrpcode', models.TextField(max_length=10)),
+                ('courseyear', models.BigIntegerField(blank=True, null=True)),
+                ('fixingrp', models.CharField(blank=True, max_length=1, null=True)),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py]: Will not change migrations manually

File: backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py [hunk 8]
```diff
@@ -1,60 +1,60 @@
...
+                ('inactive', models.CharField(blank=True, max_length=1, null=True)),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py]: Will not change migrations manually

File: backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py [hunk 9]
```diff
@@ -1,60 +1,60 @@
...
+            name='StuclassesA',
+            fields=[
+                ('id', models.AutoField(primary_key=True, serialize=False)),
+                ('setid', models.TextField(max_length=10)),
+                ('studentid', models.TextField(max_length=12)),
+                ('courseid', models.TextField(max_length=12)),
+                ('classgroupid', models.TextField(max_length=10)),
+                ('clsgrpcode', models.TextField(max_length=10)),
+                ('courseyear', models.BigIntegerField(blank=True, null=True)),
+                ('fixingrp', models.CharField(blank=True, max_length=1, null=True)),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py]: Will not change migrations manually

File: backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py [hunk 10]
```diff
@@ -1,60 +1,60 @@
...
+            name='Stuclasses',
+            fields=[
+                ('setid', models.TextField(max_length=10)),
+                ('studentid', models.TextField(max_length=12, primary_key=True, serialize=False)),
+                ('courseid', models.TextField(max_length=12)),
+                ('classgroupid', models.TextField(max_length=10)),
+                ('clsgrpcode', models.TextField(max_length=10)),
+                ('courseyear', models.BigIntegerField(blank=True, null=True)),
+                ('fixingrp', models.CharField(blank=True, max_length=1, null=True)),
+                ('inactive', models.CharField(blank=True, max_length=1, null=True)),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/timetable/migrations/0012_stuclasses_stuclassesa_stuclassesb.py]: Will not change migrations manually

File: backend/uclapi/timetable/app_helpers.py [hunk 11]
```diff
@@ -1,672 +1,672 @@
...
+                    elif module.lecturerid:
+                        event_data["module"]["lecturer"] = \
+                            _get_lecturer_details(module.lecturerid.strip())
+                    else:
+                        # This will give us 'Unknown' in all fields
+                        event_data["module"]["lecturer"] = \
+                            _lecturers_cache[None]
+
+
+                    date_str = booking.startdatetime.strftime("%Y-%m-%d")
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/timetable/app_helpers.py]: Changed in a future commit

File: backend/uclapi/oauth/urls.py [hunk 12]
```diff
@@ -1,18 +1,18 @@
...
-    url(r'deauthorise$', views.deauthorise_app),
-]
+from django.conf.urls import url
+
+from . import views
+
+urlpatterns = [
+    url(r'myapps/shibcallback$', views.myapps_shibboleth_callback),
+	url(r'myapps$', views.my_apps),
+    url(r'authorise/$', views.authorise),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/urls.py]: Changed in a future commit

File: backend/uclapi/oauth/urls.py [hunk 13]
```diff
@@ -1,18 +1,18 @@
...
-    url(r'user/studentnumber$', views.get_student_number),
-    url(r'deauthorise$', views.deauthorise_app),
-]
+from django.conf.urls import url
+
+from . import views
+
+urlpatterns = [
+    url(r'myapps/shibcallback$', views.myapps_shibboleth_callback),
+	url(r'myapps$', views.my_apps),
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/urls.py]: Changed in a future commit

File: backend/uclapi/oauth/app_helpers.py [hunk 14]
```diff
@@ -1,32 +1,32 @@
...
+    key = hexlify(os.urandom(40)).decode()
+    final = "verify" + key
+    return final
+
+
+def get_student_by_upi(upi):
+    """Returns a StudentA or StudentB object by UPI"""
+    students = StudentsA \
+               if Lock.objects.all()[0].a \
+               else StudentsB
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/app_helpers.py]: Changed in a future commit

File: backend/uclapi/oauth/app_helpers.py [hunk 15]
```diff
@@ -1,32 +1,32 @@
...
+def generate_random_verification_code():
+    key = hexlify(os.urandom(40)).decode()
+    final = "verify" + key
+    return final
+
+
+def get_student_by_upi(upi):
+    """Returns a StudentA or StudentB object by UPI"""
+    students = StudentsA \
+               if Lock.objects.all()[0].a \
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/app_helpers.py]: Changed in a future commit

File: backend/uclapi/oauth/app_helpers.py [hunk 16]
```diff
@@ -1,32 +1,32 @@
...
-        qtype2=upi_upper
-    )[0]
-    return student
+import os
+import textwrap
+from binascii import hexlify
+
+from timetable.models import Lock, StudentsA, StudentsB
+
+def generate_user_token():
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/oauth/app_helpers.py]: Changed in a future commit

File: backend/uclapi/manage.py [hunk 17]
```diff
@@ -1,28 +1,28 @@
...
+    )
+
+    try:
+        from django.core.management import execute_from_command_line
+    except ImportError:
+        # The above import may fail for some other reason. Ensure that the
+        # issue is really that Django is missing to avoid masking other
+        # exceptions on Python 2.
+        try:
+            import django
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/manage.py]: Ignored

File: backend/uclapi/dashboard/models.py [hunk 18]
```diff
@@ -1,134 +1,134 @@
...
+    def regenerate_token(self):
+        new_token = generate_api_token()
+        self.api_token = new_token
+        self.save()
+        return self.api_token
+
+    class Meta:
+        _DATABASE = 'default'
+
+class APICall(models.Model):
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/dashboard/models.py]: Changed in a future commit

File: backend/uclapi/dashboard/models.py [hunk 19]
```diff
@@ -1,134 +1,134 @@
...
-
-
-# Django signal
-@receiver(post_save, sender=App)
-def create_webhook_on_app_creation(sender, instance, created, **kwargs):
-    if created:
-        new_webhook = Webhook(app=instance)
-        new_webhook.save()
+from django.db import models
+from .app_helpers import (
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/dashboard/models.py]: Changed in a future commit

File: backend/uclapi/common/tests.py [hunk 20]
```diff
@@ -1,593 +1,593 @@
...
+from rest_framework.test import APIRequestFactory
+
+from uclapi.settings import REDIS_UCLAPI_HOST
+
+import datetime
+import json
+import redis
+import time
+
+class SecondsUntilMidnightTestCase(SimpleTestCase):
```
[the remark this answers is not in the document]
Comment (TuxedoFish) [reply backend/uclapi/common/tests.py]: Changed in a future commit

**Comment by TuxedoFish** [comment]
@anushks @hueyy: If either of you has any time this pull request is ready to merge. Should only be aesthetic changes, the tests haven't been changed but the endpoints for the index page and the about page have been modified. P.S I accidentally made HoundCI Look at SCSS now so yay? I am going to make hound ignore certain files and sort out scss

**Comment by anushks** [comment]
> I'm not sure why git thinks you've rewritten several files when there are actually no changes at all (e.g. README, CONTRIBUTING, `backend/uclapi/common/helpers.py`, etc.). Also, you seem to have unintentionally made some minor changes to several files (e.g. `backend/uclapi/common/decorators.py`, `backend/uclapi/common/tests.py`, `backend/uclapi/dashboard/app_helpers.py` where you added a line break). I'm guessing you removed the files and re-added them back at some point?
> 
> This messes up the git history a bit because `git blame` will now fault you, and it makes code reviews trickier because the diffs are not very useful.
> 
> If you're not opposed I think it would be safer to squash your commits, then we can properly diff to make sure there are no unintentional changes to the core API.


@hueyy yeah I agree, this is probably the best thing to do.
@TuxedoFish as far as I can see through these files all is well, need to sort out some of these hound errors, but yeah we should make sure we haven't upset any of the other files when we diff properly

**Comment by TuxedoFish** [comment]
@hueyy @anushks Yeah I will squash to avoid git thinking I have rewritten entire files that sounds real annoying. I intentionally added extra line breaks in for hound though! Cheers @anushks and @hueyy

**Comment by TuxedoFish** [comment]
> I'm not sure why git thinks you've rewritten several files when there are actually no changes at all (e.g. README, CONTRIBUTING, `backend/uclapi/common/helpers.py`, etc.). Also, you seem to have unintentionally made some minor changes to several files (e.g. `backend/uclapi/common/decorators.py`, `backend/uclapi/common/tests.py`, `backend/uclapi/dashboard/app_helpers.py` where you added a line break). I'm guessing you removed the files and re-added them back at some point?
> 
> This messes up the git history a bit because `git blame` will now fault you, and it makes code reviews trickier because the diffs are not very useful.
> 
> If you're not opposed I think it would be safer to squash your commits, then we can properly diff to make sure there are no unintentional changes to the core API.

Going to do a rebase today to try sort out the history, could take a while should do git rebase not git merge for the future

**Comment by TuxedoFish** [comment]
@hueyy and @anushks Just ran a rebase took way longer and was a bit tricky. Still seems to think I have changed files I never touched? Not entirely sure what to do about this because it was a pain going through the git rebase process. I think it was when I was working on the work mac it didnt have autocrlf set to true so it thinks I touched files I didnt

**Comment by TuxedoFish** [comment]
Going to try reset those specific files to master

File: .gitignore [hunk 21]
```diff
@@ -9,6 +9,10 @@ Full_DDL.sql
 # C extensions
 *.so
 
+# MERGING
```
Comment (hueyy) [inline .gitignore]: Should we still be ignoring these files?

File: .gitignore [hunk 22]
```diff
@@ -142,4 +146,5 @@ yarn.lock
 
 # django-webpack-loader
 webpack-stats.json
-backend/uclapi/src
\ No newline at end of file
+backend/uclapi/src
+frontend/webpack.dev.js
```
Comment (hueyy) [inline .gitignore]: `webpack.dev.js` should actually be committed. I think Anush temporarily put this in, but we should take it out before we merge.

File: frontend/src/components/layout/views/CodeView.jsx [hunk 23]
```diff
@@ -0,0 +1,67 @@
+// Standard React components
+import React from 'react'; 
+
+// DEPENDENCIES
+import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
+import SyntaxHighlighter from 'react-syntax-highlighter';
+import {androidstudio} from 'react-syntax-highlighter/dist/esm/styles/hljs';
+import 'whatwg-fetch';
```
Comment (hueyy) [inline frontend/src/components/layout/views/CodeView.jsx]: Why are we importing this here? We don't seem to be making any fetch requests within this file.

File: frontend/src/pages/Marketplace.jsx [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+import logo from 'Images/home-page/logo.svg';
+import arrow from 'Images/marketplace/arrow-left.svg';
+
+// Common Components
+import { Row, Column, TextView, ButtonView, CardView, ImageView, NavBar, Footer } from 'Layout/Items.jsx';
+
+// Grab titles and descriptions of app
+import { allApps } from 'Layout/data/app_pages.jsx';
+
+class Marketplace extends React.Component {
```
Comment (hueyy) [inline frontend/src/pages/Marketplace.jsx]: Why are there 2 very similar components, one named Marketplace and the other AppPage? `<Marketplace>` doesn't seem to be used?

**Comment by hueyy** [comment]
We appear to have 2 Documentation page components: [`documentation.jsx`](https://github.com/uclapi/uclapi/blob/harry/frontend-refactor-nochange/frontend/src/pages/documentation.jsx) and [`Documentation.jsx`](https://github.com/uclapi/uclapi/blob/harry/frontend-refactor-nochange/frontend/src/pages/Documentation.jsx). I think only one of them actually in use?

File: frontend/src/sass/common/uclapi.scss [hunk 25]
```diff
@@ -0,0 +1,439 @@
+/**
+	UCL API WEBSITE STYLES
```
Comment (hueyy) [inline frontend/src/sass/common/uclapi.scss]: Nice work here. Much neater!

File: getting-started-docker.txt [hunk 26]
```diff
@@ -0,0 +1,32 @@
+# Developing using docker
```
Comment (hueyy) [inline getting-started-docker.txt]: Should this be a markdown file, and should it be part of the main README.md?

File: frontend/src/sass/common/uclapi.scss [hunk 27]
```diff
@@ -0,0 +1,439 @@
...
+
+.row {
+	width: 100%;
+  	position: relative;
+	height: auto;
+	min-height: 100% !important;
+}
+
+.vertical-padding {
+  padding: 50px 0;
```
Comment (hueyy) [inline frontend/src/sass/common/uclapi.scss]: My personal opinion is that 50px is a bit much. I think 20px will do just fine.

Comment (hueyy) [reply frontend/src/sass/common/uclapi.scss]: Annotation 2019-10-17 194644
Annotation 2019-10-17 194611

Comment (hueyy) [reply frontend/src/sass/common/uclapi.scss]: Also, the bottom margin for the h5 makes the vertical padding uneven.
Annotation 2019-10-17 194934

File: frontend/src/pages/HomePage.jsx [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+                </Row>
+              </CardView> ) ) }
+          </Column>
+        </Row>
+
+        <Demo />
+
+        <Row isPadded={true} color="splash-parallax">         
+          <Column width="9-10" horizontalAlignment="center">
+            <TextView text={"Check out our blog for tutorials"} heading={1} align={"center"}/>
```
Comment (hueyy) [inline frontend/src/pages/HomePage.jsx]: We don't actually have tutorials yet tho 😅

File: frontend/src/pages/Marketplace.jsx [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+    var screenshotheight = "384px";
+
+    return (
+      <React.Fragment>
+          <NavBar isScroll={false} />
+
+          <Row height="600px" color="splash-parallax">         
+            <Column width="2-3" horizontalAlignment="center" verticalAlignment="center">
+              <TextView text={"UCL Marketplace"} heading={1} align={"center"}/>
+              <TextView text={"Apps to improve student life at UCL"} heading={2} align={"center"}/>
```
Comment (hueyy) [inline frontend/src/pages/Marketplace.jsx]: Maybe "Apps that use the UCL API" instead? We'd coceivably include apps that help staff, etc., here as well.

File: frontend/src/pages/AboutPage.jsx [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+
+  render () {
+      return (
+        <React.Fragment>
+          <NavBar isScroll={false}/>
+
+          <Row height = "600px" color="team-parallax">         
+            <Column width="2-3" horizontalAlignment="center" verticalAlignment="center">
+              <TextView text={"The Team"} heading={1} align={"center"}/>
+              <TextView text={`UCL API is a student led project founded by Wilhelm Klopp that opens up the massive amount of
```
Comment (hueyy) [inline frontend/src/pages/AboutPage.jsx]: It's a bit odd to put an explanation of the project under the header `The Team`. Maybe `About Us` instead, or drop the header completely.

Might also want to rephrase the text slightly: UCL API is a student-led platform that provides students and staff with internal data so they can develop tools to enrich their life at UCL.

Then move Wilhelm to his own section.

File: frontend/src/sass/common/uclapi.scss (hunk 27 above)
Comment (hueyy) [reply frontend/src/sass/common/uclapi.scss]: There should also be some horizontal padding

image

File: frontend/src/components/layout/NavBar.jsx [hunk 31]
```diff
@@ -1,127 +1,136 @@
...
+		link: "/dashboard",
+		src: "dashboard",
+	},
+	{
+		name: "marketplace",
+		link: "/marketplace",
+		src: "market",
+	},
+]
+const SlideDown = posed.div({
```
Comment (hueyy) [inline frontend/src/components/layout/NavBar.jsx]: I'm not sure the horizontal animation works well.

File: .gitignore (hunk 22 above)
Comment (TuxedoFish) [reply .gitignore]: Taken this out cheers!

File: .gitignore (hunk 21 above)
Comment (TuxedoFish) [reply .gitignore]: Was having issues with re base just changed this

File: frontend/src/pages/Marketplace.jsx (hunk 24 above)
Comment (TuxedoFish) [reply frontend/src/pages/Marketplace.jsx]: Marketplace is the page that shows all of the apps /marketplace and apppage is an indivudal app page /marketplace/:appid:

File: frontend/src/sass/common/uclapi.scss (hunk 25 above)
Comment (TuxedoFish) [reply frontend/src/sass/common/uclapi.scss]: Cheers!! Thanks so much for such a thorough review as well  💃

File: getting-started-docker.txt (hunk 26 above)
Comment (TuxedoFish) [reply getting-started-docker.txt]: You know what I completely forgotten I had done this, I will add this in at bottom of the main README.md

File: frontend/src/components/layout/NavBar.jsx (hunk 31 above)
Comment (TuxedoFish) [reply frontend/src/components/layout/NavBar.jsx]: Sorted

File: frontend/src/pages/Marketplace.jsx (hunk 29 above)
Comment (TuxedoFish) [reply frontend/src/pages/Marketplace.jsx]: Changed

**Pull request merged**